server/src/main/java/org/elasticsearch/features/FeatureService.java

```diff
@@ -116,8 +116,15 @@ public boolean clusterHasFeature(ClusterState state, NodeFeature feature) {
         if (state.clusterFeatures().clusterHasFeature(feature)) {
             return true;
         }
+        return clusterHasHistoricalFeature(state.getNodes().getMinNodeVersion(), feature.id());
+    }
 
-        var features = historicalFeatures.floorEntry(state.getNodes().getMinNodeVersion());
-        return features != null && features.getValue().contains(feature.id());
+    /**
+     * Returns {@code true} if nodes of the given {@code version} support feature with id {@code featureId}.
+     * Protected access for test purposes.
+     */
+    protected boolean clusterHasHistoricalFeature(Version version, String featureId) {
```

Comment:
I would change this to `NodeFeature`, and add an assert that the feature is a historical feature, as known through `historicalFeatures`

Reply:
I can surely add the assert, nice idea.
I'm a bit reluctant to make this a NodeFeature; I'm biased because I know that in tests we will not have access to the statically defined NodeFeature and will need to use their ids instead, so making this a NodeFeature parameter will mean that we have to create a bogus NodeFeature just to extract the id.

Reply:
I suppose you just need a `String` to check `historicalVersions` - so that should do

Reply:
Moving to composition over inheritance, this is not a problem anymore - no more protected method